actors/builtin/power/power_test.go

```diff
@@ -752,6 +765,161 @@ func TestSubmitPoRepForBulkVerify(t *testing.T) {
+		miner4 := tutil.NewIDAddr(t, 104)
+		powerUnit := power.ConsensusMinerMinPower
+
+		rt, ac := basicPowerSetup(t)
+		info1 := sealInfo(0)
+		info2 := sealInfo(1)
+		ac.submitPoRepForBulkVerify(rt, miner1, info1)
+		ac.submitPoRepForBulkVerify(rt, miner1, info2)
+
+		info3 := sealInfo(100)
+		info4 := sealInfo(101)
+		ac.submitPoRepForBulkVerify(rt, miner2, info3)
+		ac.submitPoRepForBulkVerify(rt, miner2, info4)
+
+		info5 := sealInfo(200)
+		info6 := sealInfo(201)
+		ac.submitPoRepForBulkVerify(rt, miner3, info5)
+		ac.submitPoRepForBulkVerify(rt, miner3, info6)
+
+		info7 := sealInfo(300)
+		info8 := sealInfo(301)
+		ac.submitPoRepForBulkVerify(rt, miner4, info7)
+		ac.submitPoRepForBulkVerify(rt, miner4, info8)
+
+		// TODO Because read order of keys in a multi-map is not as per insertion order,
+		// we have to move around the expected sends
+		cs := []confirmedSectorSend{{miner1, []int{0, 1}},
+			{miner3, []int{200, 201}},
+			{miner4, []int{300, 301}},
+			{miner2, []int{100, 101}}}
+
+		infos := map[addr.Address][]abi.SealVerifyInfo{miner1: []abi.SealVerifyInfo{*info1, *info2},
+			miner2: []abi.SealVerifyInfo{*info3, *info4},
+			miner3: []abi.SealVerifyInfo{*info5, *info6},
+			miner4: []abi.SealVerifyInfo{*info7, *info8}}
+
+		ac.createMinerBasic(rt, owner, owner, miner1)
+		ac.createMinerBasic(rt, owner, owner, miner2)
+		ac.createMinerBasic(rt, owner, owner, miner3)
+		ac.createMinerBasic(rt, owner, owner, miner4)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+
+		expectedPower := big.Mul(big.NewInt(4), powerUnit)
+
+		delta := abi.NewTokenAmount(1)
+		ac.updatePledgeTotal(rt, miner1, delta)
+		ac.onEpochTickEnd(rt, 0, expectedPower, cs, infos)
+
+		st := getState(rt)
+		require.EqualValues(t, delta, st.ThisEpochPledgeCollateral)
+		require.EqualValues(t, expectedPower, st.ThisEpochQualityAdjPower)
+		require.EqualValues(t, expectedPower, st.ThisEpochRawBytePower)
+	})
+
+	t.Run("success when no confirmed sector", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		ac.onEpochTickEnd(rt, 0, big.Zero(), nil, nil)
+	})
+
+	t.Run("fails if batch verify seals does not return result for a miner", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		miner2 := tutil.NewIDAddr(t, 201)
+		info1 := sealInfo(1)
+		info2 := sealInfo(2)
+
+		ac.submitPoRepForBulkVerify(rt, miner1, info1)
+		ac.submitPoRepForBulkVerify(rt, miner2, info2)
+
+		infos := map[addr.Address][]abi.SealVerifyInfo{miner1: []abi.SealVerifyInfo{*info1},
+			miner2: []abi.SealVerifyInfo{*info2}}
+
+		res := batchVerifyDefaultOutput(infos)
+		delete(res, miner1)
+		rt.ExpectBatchVerifySeals(infos, res, nil)
+		rt.ExpectValidateCallerAddr(builtin.CronActorAddr)
+
+		rt.SetEpoch(abi.ChainEpoch(0))
+		rt.SetCaller(builtin.CronActorAddr, builtin.CronActorCodeID)
```

Comment:
Use the existing `onEpochTickEnd` inside the `ExpectAbort`

Reply:
@anorth  The existing `onEpochTickEnd` sets "correct" assumptions about the `BatchVerifySeals` call.

```
	rt.ExpectBatchVerifySeals(infos, batchVerifyDefaultOutput(infos), nil)

```

This makes it easy for the other callers to use it. Since we want to program incorrect behaviour in this method here, I'm not sure of how to use `onEpochTickEnd ` here.